File: internal/output.go
package internal

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
)

// Output abstracts how commands emit results.
// CLI and Web are two implementations of the same interface.
type Output interface {
	// All commands
	Info(msg string)
	Result(v any)

	// Streaming (send)
	Thinking(token string)
	Text(token string)
	ToolCall(name, args string)
	ToolResult(content string)
	Inject(content string)
	Done()
}

// --- CLIOutput: human-readable, stdout + stderr ---

type CLIOutput struct {
	out io.Writer
	err io.Writer
}

func NewCLIOutput(out, err io.Writer) *CLIOutput {
	return &CLIOutput{out: out, err: err}
}

func DefaultCLIOutput() *CLIOutput {
	return NewCLIOutput(os.Stdout, os.Stderr)
}

func (o *CLIOutput) Info(msg string)       { fmt.Fprintln(o.err, msg) }
func (o *CLIOutput) Result(v any)          { json.NewEncoder(o.out).Encode(v) }
func (o *CLIOutput) Thinking(token string) { fmt.Fprint(o.err, token) }
func (o *CLIOutput) Text(token string)     { fmt.Fprint(o.out, token) }
func (o *CLIOutput) Done()                 { fmt.Fprintln(o.out) }

func (o *CLIOutput) ToolCall(name, args string) {
	fmt.Fprintf(o.err, "[tool] %s(%s)\n", name, truncate(args, 80))
}

func (o *CLIOutput) ToolResult(content string) {
	fmt.Fprintf(o.err, "  → %s\n", content)
}

func (o *CLIOutput) Inject(content string) {
	fmt.Fprintf(o.err, "[injected] %s\n", content)
}

// --- JSONLOutput: structured, all on stdout ---

type JSONLOutput struct {
	enc *json.Encoder
}

func NewJSONLOutput() *JSONLOutput {
	enc := json.NewEncoder(os.Stdout)
	enc.SetEscapeHTML(false)
	return &JSONLOutput{enc: enc}
}

func (o *JSONLOutput) emit(v any) { o.enc.Encode(v) }

func (o *JSONLOutput) Info(msg string) {
	o.emit(map[string]string{"type": "info", "message": msg})
}

func (o *JSONLOutput) Result(v any) {
	o.emit(map[string]any{"type": "result", "data": v})
}

func (o *JSONLOutput) Thinking(token string) {
	o.emit(map[string]string{"type": "thinking", "content": token})
}

func (o *JSONLOutput) Text(token string) {
	o.emit(map[string]string{"type": "text", "content": token})
}

func (o *JSONLOutput) ToolCall(name, args string) {
	o.emit(map[string]string{"type": "tool_call", "name": name, "arguments": args})
}

func (o *JSONLOutput) ToolResult(content string) {
	o.emit(map[string]string{"type": "tool_result", "content": content})
}

func (o *JSONLOutput) Inject(content string) {
	o.emit(map[string]string{"type": "inject", "content": content})
}

func (o *JSONLOutput) Done() {
	o.emit(map[string]string{"type": "done"})
}

// NewOutput creates an Output based on format string ("raw" or "jsonl").
func NewOutput(format string) Output {
	if format == "jsonl" {
		return NewJSONLOutput()
	}
	return DefaultCLIOutput()
}

// AsyncFileOutput writes to a run's output file (for background workers).
func AsyncFileOutput(runID string) *CLIOutput {
	path := runOutputPath(runID)
	f, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0o644)
	if err != nil {
		return NewCLIOutput(io.Discard, io.Discard)
	}
	return NewCLIOutput(f, f)
}

File: src/output.ts
import { createWriteStream, existsSync, mkdirSync, openSync, readFileSync, writeSync } from "node:fs";
import type { Writable } from "node:stream";
import { dirname } from "node:path";
import { runOutputPath } from "./paths";
import { truncateRunes } from "./shared";

export interface Output {
  info(message: string): void;
  result(value: unknown): void;
  thinking(token: string): void;
  text(token: string): void;
  toolCall(name: string, args: string): void;
  toolResult(content: string): void;
  inject(content: string): void;
  done(): void;
}

interface Writer {
  write(chunk: string): void;
}

class StreamWriter implements Writer {
  constructor(private readonly stream: Writable | NodeJS.WriteStream) {}

  write(chunk: string): void {
    this.stream.write(chunk);
  }
}

class SyncFileWriter implements Writer {
  private readonly fd: number;

  constructor(path: string) {
    mkdirSync(dirname(path), { recursive: true });
    this.fd = openSync(path, "a");
  }

  write(chunk: string): void {
    writeSync(this.fd, chunk);
  }
}

class CLIOutput implements Output {
  constructor(private readonly stdout: Writer, private readonly stderr: Writer) {}

  info(message: string): void {
    this.stderr.write(`${message}\n`);
  }

  result(value: unknown): void {
    this.stdout.write(`${JSON.stringify(value)}\n`);
  }

  thinking(token: string): void {
    this.stderr.write(token);
  }

  text(token: string): void {
    this.stdout.write(token);
  }

  toolCall(name: string, args: string): void {
    this.stderr.write(`[tool] ${name}(${truncateRunes(args, 80)})\n`);
  }

  toolResult(content: string): void {
    this.stderr.write(`  → ${content}\n`);
  }

  inject(content: string): void {
    this.stderr.write(`[injected] ${content}\n`);
  }

  done(): void {
    this.stdout.write("\n");
  }
}

class JSONLOutput implements Output {
  constructor(private readonly writer: Writer) {}

  private emit(value: unknown): void {
    this.writer.write(`${JSON.stringify(value)}\n`);
  }

  info(message: string): void {
    this.emit({ type: "info", message });
  }

  result(value: unknown): void {
    this.emit({ type: "result", data: value });
  }

  thinking(token: string): void {
    this.emit({ type: "thinking", content: token });
  }

  text(token: string): void {
    this.emit({ type: "text", content: token });
  }

  toolCall(name: string, args: string): void {
    this.emit({ type: "tool_call", name, arguments: args });
  }

  toolResult(content: string): void {
    this.emit({ type: "tool_result", content });
  }

  inject(content: string): void {
    this.emit({ type: "inject", content });
  }

  done(): void {
    this.emit({ type: "done" });
  }
}

export function createCLIOutput(): Output {
  return new CLIOutput(new StreamWriter(process.stdout), new StreamWriter(process.stderr));
}

export function createJSONLOutput(): Output {
  return new JSONLOutput(new StreamWriter(process.stdout));
}

export function createJSONLChunkOutput(onChunk: (chunk: string) => void): Output {
  return new JSONLOutput({ write: onChunk });
}

export function createAsyncFileOutput(runId: string): Output {
  return new CLIOutput(new SyncFileWriter(runOutputPath(runId)), new SyncFileWriter(runOutputPath(runId)));
}

export function createOutput(format = "raw"): Output {
  return format === "jsonl" ? createJSONLOutput() : createCLIOutput();
}

export function readOutput(runId: string): string {
  const path = runOutputPath(runId);
  if (!existsSync(path)) {
    return "";
  }
  return readFileSync(path, "utf8");
}
